### ISSUE TITLE
[Refactor] Create optimizer_utils.mojo for shared optimizer logic

### ISSUE THREAD

**Issue by mvillmow** [mvillmow/ml-odyssey#2467]
## Problem

Duplicated code across 5 optimizer files in `shared/training/optimizers/`:
- SGD (~150+ lines)
- Adam (~150+ lines)
- RMSprop
- AdamW
- LARS

**Duplicated patterns:**
1. **Shape validation** (~5 lines each, 5 files = 25 lines):
   ```mojo
   if params.shape() != gradients.shape():
       raise Error("Parameters and gradients must have the same shape")
   if params.dtype() != gradients.dtype():
       raise Error("Parameters and gradients must have the same dtype")
   ```

2. **Weight decay implementation** (~10 lines each, 4 files = 40 lines):
   ```mojo
   if weight_decay > 0.0:
       var wd_tensor = full_like(params, weight_decay)
       var decay_term = multiply_simd(wd_tensor, params)
       effective_gradients = add_simd(gradients, decay_term)
   ```

3. **Learning rate tensor creation** (~3 lines each, 5 files = 15 lines):
   ```mojo
   var lr_tensor = full_like(params, learning_rate)
   var update = multiply_simd(lr_tensor, effective_gradients)
   ```

## Files Affected

- `shared/training/optimizers/sgd.mojo`
- `shared/training/optimizers/adam.mojo`
- `shared/training/optimizers/rmsprop.mojo`
- `shared/training/optimizers/adamw.mojo`
- `shared/training/optimizers/lars.mojo`
- New: `shared/training/optimizers/optimizer_utils.mojo`

## Proposed Solution

1. Create `shared/training/optimizers/optimizer_utils.mojo`:
   ```mojo
   fn validate_optimizer_inputs(
       params: ExTensor,
       gradients: ExTensor
   ) raises:
       """Validate parameter and gradient tensors match."""
       if params.shape() != gradients.shape():
           raise Error("Parameters and gradients must have the same shape")
       if params.dtype() != gradients.dtype():
           raise Error("Parameters and gradients must have the same dtype")
   
   fn apply_weight_decay(
       gradients: ExTensor,
       params: ExTensor,
       weight_decay: Float64
   ) -> ExTensor:
       """Apply L2 weight decay to gradients."""
       if weight_decay > 0.0:
           var wd_tensor = full_like(params, weight_decay)
           var decay_term = multiply_simd(wd_tensor, params)
           return add_simd(gradients, decay_term)
       return gradients
   
   fn create_lr_update(
       gradients: ExTensor,
       learning_rate: Float64
   ) -> ExTensor:
       """Create learning rate scaled update."""
       var lr_tensor = full_like(gradients, learning_rate)
       return multiply_simd(lr_tensor, gradients)
   ```

2. Update optimizers to use shared utilities:
   ```mojo
   fn step(mut self, params: ExTensor, gradients: ExTensor) -> ExTensor:
       validate_optimizer_inputs(params, gradients)
       var grad = apply_weight_decay(gradients, params, self.weight_decay)
       var update = create_lr_update(grad, self.learning_rate)
       # ... optimizer-specific logic
   ```

## Expected Impact

- **400-600 lines saved (80% reduction in duplicated code)**
- Consistent optimizer behavior
- Easier to add new optimizers
- Bug fixes apply to all optimizers

## Priority

High (Phase 2)
